File: python/ReverseNodesink-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def splice(self, head, k) :
        newhead, i, travel= None, 0, head
        while i < k and travel :
            travel, i = travel.next, i+1
        if i < k : return head, None
        i=0
        while i < k and head :
            cur, head = head, head.next
            cur.next, newhead = newhead, cur
            i = i+1
        return newhead, head
    def reverseKGroup(self, head, k):
        new, old = self.splice(head, k)
        last  = head
        while old :
            n1, o1 = self.splice(old, k)
            last.next = n1
            last, old = old, o1
        return new

File: cpluplus/Reverse_Nodes_in_k-Group.cc
/*
 * author: yuandx
 * create: 2015-11-22
 * email: yuandx@mvad.com
 */

#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <climits>
#include <string>
#include <utility>
#include <functional>
#include <algorithm>
#include <stdio.h>
#include <sstream>
#include <unordered_map>
#include <unordered_set>

using std::unordered_map;
using std::unordered_set;
using std::ostringstream;
using std::istringstream;
using std::stringstream;
using std::cout;
using std::cin;
using std::endl;
using std::vector;
using std::map;
using std::set;
using std::pair;
using std::sort;
using std::string;
using std::for_each;
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};
class Solution
{
public:
  ListNode* reverse(ListNode* head)
  {
    ListNode* new_head = NULL;
    while (head != NULL)
    {
      ListNode* next = head->next;
      head->next = new_head;
      new_head = head;
      head = next;
    }
    return new_head;
  }

  ListNode* reverseKGroup(ListNode* head, int k)
  {
    ListNode tmp(0);
    tmp.next = head; head = &tmp;
    ListNode* prev = &tmp;
    while (prev != NULL)
    {
      while(count < k && cur != NULL)
      {
        ListNode* next = cur->next;
        cur->next = cur_head;
        cur_head = cur;
        count++;
        cur = next;
      }
      ListNode* next_prev = prev->next;
      prev->next = cur_head;
      if (next_prev != NULL)
      {
        next_prev->next = cur;
      }
      prev = next_prev;
    }
    return head->next;
  }
};

int main()
{

  return 0;
}
